hypothesis-python/src/hypothesis/searchstrategy/collections.py

```diff
@@ -143,13 +147,18 @@ def do_draw(self, data):
         seen = set()
         result = []
 
+        # We construct a filtered strategy here rather than using a check-and-reject
+        # approach because some strategies have special logic for generation under a
+        # filter, and FilteredStrategy can consolidate multiple filters.
+        filtered = self.element_strategy.filter(lambda val: self.key(val) not in seen)
         while elements.more():
-            value = data.draw(self.element_strategy)
-            k = self.key(value)
-            if k in seen:
+            value = self.element_strategy.do_filtered_draw(
```

Comment:
I think I see what you're doing here, but I'm not keen on the idea of other strategies manually calling `do_filtered_draw` outside of `FilteredStrategy` itself.

(I understand that we can't just call `filtered.do_draw(data)`, because a failed draw would cancel the whole top-level example instead of just signalling to `UniqueListStrategy` that it should stop with what it already has.)

Reply:
Happy to take suggestions here - the previous hand-implementation for `sampled_from` was even more efficient, and if you don't mind the special casing (I don't) it could also unwrap filtered `sampled_from`s...